### ISSUE TITLE
Add OIDC Provider Support

### ISSUE THREAD

**Issue by richardcase** [kubernetes-sigs/cluster-api-provider-aws#2248]
/kind feature
/area provider/eks

**Describe the solution you'd like**
You can associate an OIDC provider with an EKS cluster and use that to authenticate users who access the cluster. We should add support for this to CAPA

**Anything else you would like to add:**
See the announcement [here](https://aws.amazon.com/blogs/containers/introducing-oidc-identity-provider-authentication-amazon-eks/).

**Environment:**

- Cluster-api-provider-aws version: 0.6.4

**Comment by enxebre** [comment]
What is preventing this from happening today?

Couldn't a cluster admin setup an OIDC provider and the right kube api server flags:

```
--service-account-issuer
--service-account-key-file
--service-account-signing-key-file
--api-audiences
```

Then create the right iam roles with appropriate `sts:AssumeRole` policies. And then the aws creds secret as

```
role_arn = %s
web_identity_token_file = /var/run/secrets/serviceaccount/token
```

**Comment by randomvariable** [comment]
I think this is an EKS specific thing, although that functionally translates into the same API server setting that's hidden inside their black box.

**Comment by richardcase** [comment]
Thats right

**Comment by sadysnaat** [comment]
Hi guys we have a requirement for this feature, I would like to take this up 

We propose to add this field to AWSManagedControlPlaneSpec

```
type AWSManagedControlPlaneSpec struct {

…..

OidcIdentityProviderConfigRequest *OidcIdentityProviderConfigRequest

}
```


### Behaviour 

setting a value will disassociate any other IdentityProvider and register the IdentityProvider from AWSManagedControlPlaneSpec 

Setting this to nil will dissociate any preregistered IdentityProvider 

Changing the values in OidcIdentityProviderConfigRequest will deregister the old IdentityProvider and register the new ones

**Comment by richardcase** [comment]
@sadysnaat - i had started to work on this but didn't very far at all. I had started to make the API changes: 

https://github.com/richardcase/cluster-api-provider-aws/commits/2248-oidc-provider

If you wanted to copy the definitions and apply to v1alpha4 AWSManagedControlPlane instead that would be good.  

If you need any help with the change feel free to ping me here or on slack.

/unassign

**Comment by richardcase** [comment]
Fixed by #2422
